stdlib/json.gr

```diff
@@ -0,0 +1,2308 @@
+    JSONObject(entries) => {
+      match (entries) {
+        [] => {
+          Buffer.addChar('{', buffer)
+          Buffer.addChar('}', buffer)
+          None
+        },
+        [(key, value)] => {
+          Buffer.addChar('{', buffer)
+
+          implHelper.emitEscapedQuotedString(key)
+
+          Buffer.addChar(':', buffer)
+
+          let err = printElementCompact(value, implHelper)
+
+          if (Option.isNone(err)) {
+            Buffer.addChar('}', buffer)
+          }
+
+          err
+        },
+        [initialHead, ...initialRest] => {
+          Buffer.addChar('{', buffer)
+
+          let mut currentHead = initialHead
+          let mut currentRest = initialRest
+
+          let mut err = None
+
+          for (let mut index = 0; ; index += 1) {
+            if (index > 0) {
+              Buffer.addChar(',', buffer)
+            }
+
+            let (key, value) = currentHead
+
+            implHelper.emitEscapedQuotedString(key)
+
+            Buffer.addChar(':', buffer)
+
+            err = printElementCompact(value, implHelper)
+
+            if (Option.isSome(err)) {
+              break
+            }
+
+            match (currentRest) {
+              [] => break,
+              [newHead, ...newRest] => {
+                currentHead = newHead
+                currentRest = newRest
+              },
+            }
+          }
+
+          if (Option.isNone(err)) {
+            Buffer.addChar('}', buffer)
+          }
+
+          err
+        },
+      }
+    },
+  }
+}
+
+let makeCompactJSONWriter = (buffer: Buffer.Buffer) => {
+  // As an optimization prepare a closure function to execute for each code
+  // point of the input. It's important that this is not inlide in the
+  // emitEscapedQuotedString function directly to avoid allocating the closue
+  // for each input string.
+  let emitCodePoint = (codePoint: Number) => {
+    // For the compact formatting only escape backslash, double quotes and
+    // code points 0..31 (the so called "C0" control point group) as
+    // required by ECMA-404. For the non pretty printed case we don't want
+    // to escape more than what is stricty required to avoid increasing the
+    // output size.
+    if (codePoint > 31 && codePoint != 0x0022 && codePoint != 0x005C) {
+      addCharFromCodePoint(codePoint, buffer)
+    } else {
+      emitEscapedCodePoint(codePoint, buffer)
+    }
+  }
+
+  let emitEscapedQuotedString = (s: String) => {
+    Buffer.addChar('"', buffer)
+
+    String.forEachCodePoint(emitCodePoint, s)
+
+    Buffer.addChar('"', buffer)
+  }
+
+  let implHelper = {
+    buffer,
+    emitEscapedQuotedString: s => {
+      emitEscapedQuotedString(s)
+    },
+  }: JSONWriterCompactImplHelper
+
+  {
+    emit: json => {
+      let error = printElementCompact(json, implHelper)
+
+      error
+    },
+  }: JSONWriter
+}
+
+// Note that this could relatively easily be integrated with
+// printElementCompact to avoid code duplication, but it would mean
```

Comment:
I think I would advocate for this integration unless it would be an inordinate slowdown for the compact case.

Reply:
In my limited testing the compact version is about 2-4x faster, depending on input. The more nested and complex the JSON, the grater the difference.

Reply:
I agree with philip that we should create the singular `JsonWriter` and the fields can be `Option` on whether there is a function implementation to keep the performance of 2 impls

Reply:
so merge `JsonWriterPrettyImplHelper` and `JsonWriterCompactImplHelper` make the functions that are part of `PrettyImplHelper` options? and then merge merge `makeCompactJsonWriter` and `makePrettyPrintJsonWriter`?

Reply:
I merged the writers, I found the best way todo it was just to get rid of the compact one and pass the formatting options for compact to the pretty one, as adding options would have probably been slower.